chore(style): Profile styling tweaks and cleanup

the-j0k3r opened the pull request:
I have a few more things todo here and some things are somewhat weird.

# TODO: 

- [x] Fix input lines being cut. https://github.com/bus-stop/x-terminal/pull/16#issuecomment-593362409
- [x] Fix alignments
- [x] Tweak wording
- [x] Fix buttons
- [x] Add octicons to buttons and dialogs
- [x] Fix checkbox styling

### Before and after

![generic](https://user-images.githubusercontent.com/31389848/75629189-10bf5d00-5be0-11ea-961a-765df2d76f59.gif)

<s>## Issue 2 - checkbox background

@UziTech Any ideas about the checkbox + states styling?

https://github.com/atom/atom/blob/a69a55e92295d8285d9161e0d57b37970b898dde/packages/one-dark-ui/styles/settings.less#L125-L135

Ive tried these properties and while inspector shows this loaded the styling is unchanged.

See gif:

![generic](https://user-images.githubusercontent.com/31389848/75629270-c5597e80-5be0-11ea-8564-65e8888ef6a0.gif)

I can filter invert the checkbox making it dark but it looks just as ugly and out of place.</s>

the-j0k3r committed:
chore(style): add variables

the-j0k3r committed:
chore(style): restyle buttons + add hover color

the-j0k3r committed:
chore(style): consume variables

the-j0k3r committed:
chore(style): style hline

Atom style hr/hline

the-j0k3r commented:
@UziTech OK aside from being unable tos style the checkbox there are some **height** issues with the profile inputs.

We have at least two input that are cut off at the bottom.

![generic](https://user-images.githubusercontent.com/31389848/75673168-9bf12f00-5c82-11ea-8549-95b019151d59.gif)

Temporarily we can RE-define
```css
atom-text-editor[mini] {
line-height: 2em;
}
```

Its the same value that Atom uses, but clearly its not being loaded here and the result is ugly.

![generic](https://user-images.githubusercontent.com/31389848/75673514-69940180-5c83-11ea-86a8-aa1b177e65d1.gif)


Im not a fan of rendering stuff to canvas anyway,

UziTech commented:
You can style the check boxes by using class `input-checkbox`

```
<input class="native-key-bindings input-checkbox" type="checkbox" />
```

the-j0k3r commented:
> You can style the check boxes by using class `input-checkbox`
> 
> ```
> <input className="native-key-bindings input-checkbox" type="checkbox" />
> ```

Thats part of Atom and already defined, we shouldn't have to redefine rules.

the-j0k3r commented:
Yep @UziTech doesnt work
added

```css
.input-checkbox {
	background-color: @background-color-selected;
	&:active,
	&:checked {
		background-color: @accent-color;
	}
	&:before,
	&:after {
		background-color: @accent-text-color;
	}
}
```

Doesnt work.

UziTech commented:
Atom already has styles for `.input-checkbox` just adding the class to the input works for me.

![inpu-checkbox](https://user-images.githubusercontent.com/97994/75687325-4194a600-5c63-11ea-94bf-e0eb545074cd.gif)

the-j0k3r commented:
> Atom already has styles for `.input-checkbox` just adding the class to the input works for me.

I know, but it doesnt work when we add them to the stylesheet.

the-j0k3r commented:
Well I found a way, but the internal class name be dammed. https://github.com/bus-stop/x-terminal/pull/16/commits/7327274bc77293c636c34cde139d2baedbfe990d

I guess most of the styling failures are due to the classes used, but that should not be the case.

the-j0k3r committed:
fix: style input checkbox in profiles

x-terminal-profile-menu-item-checkbox is no more

UziTech commented:
> I know, but it doesnt work when we add them to the stylesheet.

Why do we need to add them to the style sheet if the styles are already in Atom's stylesheet?

the-j0k3r commented:
> Why do we need to add them to the style sheet if the styles are already in Atom's stylesheet?

See https://github.com/bus-stop/x-terminal/commit/7327274bc77293c636c34cde139d2baedbfe990d

Because we are using non-standard classes it wont work by simply styling our weird selectors.

looks at the old one https://github.com/bus-stop/x-terminal/blob/369c41aaca006459b7d6586852fece1d78e45fa3/styles/x-terminal.less#L115 adding properties that should work but doesnt.

I suppose it doesnt because part of the terminal is rendered in a canvas, if not that idk

UziTech commented:
we don't need to change the stylesheet at all. Just add the class `input-checkbox` to the checkbox element.

the-j0k3r commented:
> we don't need to change the stylesheet at all. Just add the class `input-checkbox` to the checkbox element.

It wont work but OK. try it, IDK JS but CSS I know quite well.

UziTech commented:
![image](https://user-images.githubusercontent.com/97994/75689095-40b14380-5c66-11ea-847c-89bb816d764c.png)

works for me

UziTech commented:
![image](https://user-images.githubusercontent.com/97994/75689218-6c342e00-5c66-11ea-9a85-631db5a5bc0c.png)

the-j0k3r commented:
You need to change both files else you get

![Capture](https://user-images.githubusercontent.com/31389848/75689652-d0c0ae80-5ca1-11ea-99d0-3b4d84d8bf6a.PNG)

UziTech commented:
mine looks fine

![image](https://user-images.githubusercontent.com/97994/75689971-bb2e9300-5c67-11ea-9926-67464134d409.png)

the-j0k3r commented:
Anywho, thats solved, what about https://github.com/bus-stop/x-terminal/pull/16#issuecomment-593362409

UziTech commented:
None of mine are cut off. Could it be something with your theme?

the-j0k3r commented:
Im using One Dark thoughout

the-j0k3r commented:
Only 1 is cut off and 1 is partially cut off, both at the bottom.

The majority are fine

the-j0k3r commented:
Its not the font (I defined same as Atom "Menlo, Consolas, DejaVu Sans Mono, monospace"
Its not the font size, default 14px I set 16px

Reverted to default and same.

I guess I could try a fresh Atom profile, I dont have many non standard packages.

UziTech commented:
Another thing that seems to need fixing is the tab order

the-j0k3r commented:
> Another thing that seems to need fixing is the tab order

Well one of the many things, mind opening a new issue for that?

UziTech commented:
I think I see what you mean by cut off

![image](https://user-images.githubusercontent.com/97994/75693839-30509700-5c6d-11ea-8adf-78af0995cd7e.png)

maybe an `overflow: hidden` somewhere?

the-j0k3r commented:
I think https://github.com/bus-stop/x-terminal/pull/16/commits/6576e9e104e31f22e356e40e19f7f7b6d72df537 fixes the input cut off issue.

![generic](https://user-images.githubusercontent.com/31389848/75698091-564a5b80-5cae-11ea-82d7-30f4ecf02ad7.gif)

OK there?

this also works with overflow: hidden and neither seem to have any effect on other portions of the terminal. if something is broken by this I havent seen it.

the-j0k3r committed:
fix input bottom cutoff when focused

Some of these were being mangled.

the-j0k3r committed:
chore(style): modal-message tweaks

* sets more or less Atoms font size
* use octicons to decorate :before

the-j0k3r committed:
chore(style): profiles main increase padding

we could go as much as * 10, but * 5 adds a padding similar to Atoms 
settings

also the whole profile menu looks better with * 5

the-j0k3r committed:
chore(fix): dont cut off checkbox left border

the-j0k3r committed:
chore(style): add buttons tab navigation

like this you can tab nav with button feedback like Atom

The inputs all need to use Atom selectors as is we cant do the same to 
those.

the-j0k3r committed:
chore(style): term link to accent-color

the-j0k3r committed:
chore(style): add border-radius to select

ala Atom

the-j0k3r committed:
chore(style): add color swatch tab navigation

Atom doesnt support tab nav in inputs and color swatches.

We add it here because its the right design choice. A user should have 
visual feedback while tab naving to know where to stop tabbing

the-j0k3r committed:
chore(style): color and improve link display

* shorten the huge properties title + link lenght

* colors link in settings same as Atom

the-j0k3r committed:
chore(style): p line height is inherited

Atom p definition reaches this and any propeties to that effect are 
overwriten.

the-j0k3r committed:
chore(style): fix alignment

the-j0k3r commented:
@UziTech Still a couple of things todo but that shouldn't affect anything else visually. :tm: famous last words.

Mind giving this a whirl in yours as is?

UziTech commented on src/profile-menu-element.js:
```diff
@@ -365,7 +365,7 @@ class XTerminalProfileMenuElementImpl extends HTMLElement {
 		menuItemLabel.classList.add('x-terminal-profile-menu-item-label-checkbox')
 		const checkbox = document.createElement('input')
 		checkbox.setAttribute('type', 'checkbox')
-		checkbox.classList.add('x-terminal-profile-menu-item-checkbox')
```
I think we should keep this class and use it in the stylesheet otherwise we might accidentally style all `.input-checkbox` elements in Atom

the-j0k3r replied:
It has to be like this see https://github.com/bus-stop/x-terminal/pull/16#discussion_r387074110 if not it only doesn not work fully but it beaks, so both have to be styled.

Ive not managed to do it like you say.

UziTech commented on src/config.js:
```diff
@@ -738,7 +738,7 @@ export const config = configOrder({
+				description: 'Options to apply to xterm.js Terminal objects. Read more on the supported [xterm.js API properties](https://xtermjs.org/docs/api/terminal/interfaces/iterminaloptions/#properties).',
 				type: 'string',
 				default: configDefaults.xtermOptions,
 				profileData: {
```
Another style thing below this, all of the references to `xTerm` should probably be changed to `xterm.js`

the-j0k3r replied:
I guess I can include those in this also.

the-j0k3r replied:
all done here I think

UziTech commented on styles/x-terminal.less:
```diff
@@ -106,16 +114,17 @@ x-terminal-profile {
-		width: 15px;
-		height: 15px;
-		margin-left: -2.25em;
+	.input-checkbox {
```
we should style class `x-terminal-profile-menu-item-checkbox` as mentioned above.

the-j0k3r replied:
It only works here like this. Anything else breaks. The important thing is it works fully.

Also other classes names will be refactored after, In most of these cases adding x-terminal to sub classes is useless.

UziTech replied:
Why would it not work the same if we add the classname `x-terminal-profile-menu-item-checkbox` and style that class instead of `input-checkbox`?

If we are running into precedence issues you can use `!important` to override the styling from `input-checkbox`.

> adding x-terminal to sub classes is useless.

It is future-proofing incase someone decides they shouldn't be nested and breaks other elements with the same class names. Or some other package decides to give their element the same class and not nest their styles.

the-j0k3r replied:
> If we are running into precedence issues you can use `!important` to override the styling from `input-checkbox`.

And you think I didnt try that?

> It is future-proofing incase someone decides they shouldn't be nested and breaks other elements with the same class names. Or some other package decides to give their element the same class and not nest their styles.

We can account for many things in the design, I never heard of such issues personally with Atom addons and forks, The way things are loaded with unique paths it would be insane of Atom to suddenly wack everything in a design that would cause these things to happen.

If it would happen, it would be an Electron/Atom issue. Besides it can be easily undone or redone its just CSS.

UziTech replied:
Do the check boxes look different with the fixes I pushed?

> Besides it can be easily undone or redone its just CSS.

Yes but it will take some time to figure out where the issue is and why it is happening. It doesn't hurt anything to have classes prefixed with`x-terminal`

the-j0k3r replied:
I still think the lot as is is overkill for the sake of unlikely situations.

UziTech reviewed (changes requested):
We should also add some space between the buttons

```diff
	createButton () {
		const button = document.createElement('button')
		button.classList.add('x-terminal-profile-menu-button')
+		button.classList.add('btn')
+		button.classList.add('inline-block-tight')
		return button
	}
```

UziTech committed:
chore(style): my fixes

UziTech commented:
See if these fixes work for you

the-j0k3r commented:
> See if these fixes work for you

It does, Ive added more stuff to other area based on that, let me push them.

the-j0k3r committed:
chore(style): add buttons + icons ala Atom

the-j0k3r committed:
chore(config.js): typos and wording tweaks

the-j0k3r commented:
@UziTech there have a test =)

the-j0k3r commented:
I will abandon the idea of refactoring class names, so this PR is good to go as is.

Please dont squash these on merge, its easier for me to keep  history and easier to look at blames

the-j0k3r committed:
chore(fix): define undefined vars

When using Atom Dark/light it will complain about these.

So lets just add them so we fully support those old themes with these 
new changes.

the-j0k3r commented:
@UziTech this now fully works with all built in Atom themes. due to https://github.com/bus-stop/x-terminal/pull/16/commits/d59ee35d0a86d9c20c038a402b3539e2449bcec8

UziTech commented:
It looks good to me. I'll let you merge it.

I added a fix for tabbing through the menu. <kbd>Shift</kbd> + <kbd>tab</kbd> doesn't work on an `atom-text-editor` but that seems to be an issue in Atom.

UziTech commented:
Just noticed, shouldn't the save icon be a floppy disc?

the-j0k3r commented:
> Just noticed, shouldn't the save icon be a floppy disc?

Is there a floppy disk in octicons? https://octicons.github.com/ ... doesnt seem so.
I did think about it, but theres none. IDK what better icon to use from the available.

Suggestions?

>It looks good to me. I'll let you merge it.

mmmk?!

UziTech commented:
> Is there a floppy disk in octicons? https://octicons.github.com/ ... doesnt seem so.

I guess not, that seems odd. Oh well.

the-j0k3r commented:
This looks like an option https://octicons.github.com/icon/dependent/

the-j0k3r commented:
Also I have a couple of fixes for Atom https://github.com/atom/settings-view , Ill submit a PR but hers them

```css
/* fixes alignment of color swatch to label
 * fix causes setting-description to shift down fixed below */ 
.settings-view {
  input[type="color"] {
    margin: 0.313em 0 0 -5em;
  }
}
/* fixes alignment setting-description to label above */ 
.color {
  .setting-description {
    margin-top: -0.5em;
  }
}
/* fixes alignment of checkbox to label and to checkmark */
.settings-view input[type="checkbox"] {
  margin: 0.29em 0.75em 0 -2.2em;
}
```

![generic](https://user-images.githubusercontent.com/31389848/75893709-84a57380-5e33-11ea-8c8e-cbbd90a75c5a.gif)

Works for all themes fixes actual Atom alignment issues with those controls

UziTech commented:
> This looks like an option octicons.github.com/icon/dependent

I think it looks good with file

the-j0k3r committed:
chore(style): style modal buttons

by default the overwirte or delete buttons are red on hover because 
these are destructive actions.

For better usability and avoid errors I would swap the button order, so 
the destructive option is further away from user.

Also consider enabling cancel button by default forcing user to actually 
select the destructive option.

the-j0k3r committed:
chore(typos): getTitle() X Terminal

the-j0k3r committed:
chore(scripts): add lintfix + rules

the-j0k3r commented:
After a cursive test, this works across OS,so merging.
